### PR TITLE
Unique data source per job

### PR DESCRIPTION
**Pull request by galalen**
## Description
Add jobId to JobRepository to create local assets uniquely for each SyftJob. [Issue 215](https://github.com/OpenMined/KotlinSyft/issues/215).

## Affected Dependencies
None.

## How has this been tested?
- Updated the test cases for JobRepository to include jodId.
## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: syft/src/main/java/org/openmined/syft/domain/JobRepository.kt [hunk 1]
```diff
@@ -199,6 +200,11 @@ internal class JobRepository(
                 }
                 .compose(config.networkingSchedulers.applySingleSchedulers())
     }
+
+    fun getLocalModelsPath(config: SyftConfiguration) = "${config.filesDir}/$jobId/models"
```
Comment (vkkhare) [inline syft/src/main/java/org/openmined/syft/domain/JobRepository.kt:204]: Better to create a new class JobDataSource that stores jobId, config.filesdir, file paths and these functions. Feed this datasource to Kob repository

File: syft/src/main/java/org/openmined/syft/execution/SyftJob.kt [hunk 2]
```diff
@@ -66,7 +66,8 @@ class SyftJob internal constructor(
                 config,
                 JobRepository(
                     JobLocalDataSource(),
-                    JobRemoteDataSource(config.getDownloader())
+                    JobRemoteDataSource(config.getDownloader()),
+                    "$modelName-$version"
```
Comment (vkkhare) [inline syft/src/main/java/org/openmined/syft/execution/SyftJob.kt:70]: JobId will later be also used for supplying request id parameters so we must ensure it is always unique. Better would be to create a class JobDataModel this will store the a unique 128bit hash as it's id, config, model name, model version. This datamodel will be fed to JobDataSource for usage.

**Review by vkkhare — CHANGES_REQUESTED** [review]

File: syft/src/main/java/org/openmined/syft/execution/SyftJob.kt (hunk 2 above)
Comment (galalen) [reply syft/src/main/java/org/openmined/syft/execution/SyftJob.kt:70]: There is an internal JobID class, I guess with this approach we don't need it anymore

File: syft/src/main/java/org/openmined/syft/domain/JobRepository.kt [hunk 3]
```diff
@@ -19,25 +18,22 @@ internal const val PLAN_OP_TYPE = "torchscript"
...
+            dataSource.localDataSource.getDiffScript(config)
 
     fun persistToLocalStorage(
         input: InputStream,
         parentDir: String,
         fileName: String,
         overwrite: Boolean = false
     ): String {
-        return jobLocalDataSource.save(input, parentDir, fileName, overwrite)
+        return dataSource.localDataSource.save(input, parentDir, fileName, overwrite)
```
Comment (vkkhare) [inline syft/src/main/java/org/openmined/syft/domain/JobRepository.kt:36]: Always call repository.save repository will handle how to use datasource. Datasource objects should not be needed outside repository

File: syft/src/main/java/org/openmined/syft/datasource/JobDataSource.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package org.openmined.syft.datasource
+
+import org.openmined.syft.datamodels.JobDataModel
+import org.openmined.syft.domain.SyftConfiguration
+
+@ExperimentalUnsignedTypes
+internal class JobDataSource(
+    val dataModel: JobDataModel,
+    val localDataSource: JobLocalDataSource,
```
[the remark this answers is not in the document]
Comment (galalen) [reply syft/src/main/java/org/openmined/syft/datasource/JobDataSource.kt:9]: because it used by the repository to download plans/models ...

File: syft/src/main/java/org/openmined/syft/domain/JobRepository.kt (hunk 3 above)
Comment (galalen) [reply syft/src/main/java/org/openmined/syft/domain/JobRepository.kt:36]: you mean to encapsulate the logic of localDataSource inside the JobDataSource? and serve the JobDataSource as a repository for local data access?

Comment (vkkhare) [reply syft/src/main/java/org/openmined/syft/domain/JobRepository.kt:36]: Nope I mean remove the jobdatasource. Have localdatasource and remote data source only. Let repository decide which one to call. A jobdatasource is redundant

File: syft/src/main/java/org/openmined/syft/datasource/JobLocalDataSource.kt [hunk 5]
```diff
@@ -114,4 +114,14 @@ internal class JobLocalDataSource {
         }
         return file.absolutePath
     }
+
+    @ExperimentalUnsignedTypes
+    fun getModelsPath(config: SyftConfiguration) = "${config.filesDir}/${dataModel.id}/models"
```
Comment (vkkhare) [inline syft/src/main/java/org/openmined/syft/datasource/JobLocalDataSource.kt:119]: I think it is better to return the complete path here. Give model id as an argument and return the complete path

File: syft/src/main/java/org/openmined/syft/datasource/JobLocalDataSource.kt [hunk 6]
```diff
@@ -11,7 +12,7 @@ import java.io.InputStream
 internal const val DIFF_SCRIPT_NAME = "diff_script.pt"
 private const val TAG = "JobLocalDataSource"
 
-internal class JobLocalDataSource {
+internal class JobLocalDataSource(@VisibleForTesting internal val dataModel: JobDataModel) {
```
Comment (vkkhare) [inline syft/src/main/java/org/openmined/syft/datasource/JobLocalDataSource.kt:15]: no need for internal specification here I guess. Also visible for testing is also default

File: syft/src/main/java/org/openmined/syft/domain/JobRepository.kt [hunk 7]
```diff
@@ -20,7 +20,7 @@ private const val TAG = "JobDownloader"
 
 @ExperimentalUnsignedTypes
 internal class JobRepository(
-    private val jobLocalDataSource: JobLocalDataSource,
+    val jobLocalDataSource: JobLocalDataSource,
     private val jobRemoteDataSource: JobRemoteDataSource
```
Comment (vkkhare) [inline syft/src/main/java/org/openmined/syft/domain/JobRepository.kt:24]: why did you remove private?

Comment (galalen) [reply syft/src/main/java/org/openmined/syft/domain/JobRepository.kt:24]: > why did you remove private?

Because I needed to access models path on SyftJob createDiff method

File: syft/src/main/java/org/openmined/syft/datamodels/JobDataModel.kt [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    val id: String get() = hashId()
+
+    /**
+     * Check if two [JobDataModel] are same. Matches both model names and version if [version] is not null for param and current jobId.
+     * @param modelName the modelName of the jobId which has to be compared with the current object
+     * @param version the version of the jobID which ahs to be compared with the current jobId
+     * @return true if JobId match
+     * @return false otherwise
+     */
+    fun matchWithResponse(modelName: String, version: String? = null) =
```
Comment (mccorby) [inline syft/src/main/java/org/openmined/syft/datamodels/JobDataModel.kt:27]: Where is this used?

File: syft/src/main/java/org/openmined/syft/domain/JobRepository.kt [hunk 9]
```diff
@@ -136,7 +141,7 @@ internal class JobRepository(
                 .flatMap { modelInputStream ->
                     jobLocalDataSource.saveAsync(
                         modelInputStream,
-                        "${config.filesDir}/models",
+                        jobLocalDataSource.getModelsPath(config),
```
Comment (mccorby) [inline syft/src/main/java/org/openmined/syft/domain/JobRepository.kt:144]: This info is already in the data source. The repository should not need to pass it again.

File: syft/src/main/java/org/openmined/syft/datamodels/JobDataModel.kt [hunk 10]
```diff
@@ -0,0 +1,42 @@
+package org.openmined.syft.datamodels
+
+import java.security.MessageDigest
+
+/**
+ * A uniquer identifier class for the job
+ * @property modelName The name of the model used in the job while querying PyGrid
+ * @property version The model version in PyGrid
+ */
+class JobDataModel(
```
Comment (mccorby) [inline syft/src/main/java/org/openmined/syft/datamodels/JobDataModel.kt:10]: This class should have a name describing its purpose. What about `JobId`?

Comment (mccorby) [reply syft/src/main/java/org/openmined/syft/datamodels/JobDataModel.kt:10]: Also, consider making it a `data class`

**Review by mccorby — DISMISSED** [review]

File: syft/src/main/java/org/openmined/syft/datamodels/JobDataModel.kt (hunk 10 above)
Comment (galalen) [reply syft/src/main/java/org/openmined/syft/datamodels/JobDataModel.kt:10]: > This class should have a name describing its purpose. What about `JobId`?

There is an inner JobId class inside SyftJob, I guess I should remove It and make this one right?

Comment (mccorby) [reply syft/src/main/java/org/openmined/syft/datamodels/JobDataModel.kt:10]: That would be good.

File: syft/src/main/java/org/openmined/syft/domain/JobRepository.kt [hunk 11]
```diff
@@ -28,6 +28,11 @@ internal class JobRepository(
     val status: DownloadStatus
         get() = trainingParamsStatus.get()
 
+    fun getLocalDataSource() = jobLocalDataSource
```
Comment (vkkhare) [inline syft/src/main/java/org/openmined/syft/domain/JobRepository.kt:31]: Why is this needed? jobLocalDataSource should be private to repository

File: syft/src/main/java/org/openmined/syft/domain/JobRepository.kt (hunk 9 above)
Comment (vkkhare) [reply syft/src/main/java/org/openmined/syft/domain/JobRepository.kt:144]: Yeah better would be to pass the config object and store it alongside for future use

File: syft/src/main/java/org/openmined/syft/execution/SyftJob.kt [hunk 12]
```diff
@@ -65,14 +65,19 @@ class SyftJob internal constructor(
...
+                        )
+                    ),
                     JobRemoteDataSource(config.getDownloader())
                 )
             )
         }
     }
 
-    val jobId = JobID(modelName, version)
+    val jobId = JobId(modelName, version)
```
Comment (vkkhare) [inline syft/src/main/java/org/openmined/syft/execution/SyftJob.kt:80]: This job ID should be passed in the repository above. Redundant data holders

**Review by vkkhare — CHANGES_REQUESTED** [review]

File: syft/src/main/java/org/openmined/syft/datasource/JobLocalDataSource.kt [hunk 13]
```diff
@@ -114,4 +114,14 @@ internal class JobLocalDataSource {
         }
         return file.absolutePath
     }
+
+    @ExperimentalUnsignedTypes
+    fun getModelsPath(config: SyftConfiguration, jobId: String) = "${config.filesDir}/$jobId/models"
```
Comment (vkkhare) [inline syft/src/main/java/org/openmined/syft/datasource/JobLocalDataSource.kt:119]: the syft repository gets the config while creation. Pass it on to the job model and store it there. So you won't need to pass syft config anymore in these methods. THen finally remove config operand from the method calls in data downloaders

File: syft/src/main/java/org/openmined/syft/domain/JobRepository.kt [hunk 14]
```diff
@@ -19,16 +17,23 @@ private const val TAG = "JobRepository"
 @ExperimentalCoroutinesApi
 @ExperimentalUnsignedTypes
 internal class JobRepository(
+    private val jobId: JobId,
     private val jobLocalDataSource: JobLocalDataSource,
-    private val jobRemoteDataSource: JobRemoteDataSource
+    private val jobRemoteDataSource: JobRemoteDataSource,
+    private val config: SyftConfiguration
```
Comment (vkkhare) [inline syft/src/main/java/org/openmined/syft/domain/JobRepository.kt:23]: Why is this config needed in repository?

File: syft/src/main/java/org/openmined/syft/execution/SyftJob.kt [hunk 15]
```diff
@@ -71,14 +71,15 @@ class SyftJob internal constructor(
                 worker,
                 config,
                 JobRepository(
-                    JobLocalDataSource(),
-                    JobRemoteDataSource(config.getDownloader())
+                    JobId(modelName, version),
```
Comment (vkkhare) [inline syft/src/main/java/org/openmined/syft/execution/SyftJob.kt:74]: Better to create a static factory method to instantiate all these job local/remote data sources from the config. Use that method everywhere

**Review by vkkhare — APPROVED** [review]
